paychmgr/simple.go

```diff
@@ -41,106 +250,268 @@ func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, am
+	smsg, err := ca.api.MpoolPushMessage(ctx, msg)
 	if err != nil {
-		return cid.Undef, err
+		return nil, err
 	}
 	mcid := smsg.Cid()
-	go pm.waitForAddFundsMsg(ctx, mcid)
-	return mcid, nil
+
+	// Store the add funds message CID on the channel
+	ca.mutateChannelInfo(from, to, func(ci *ChannelInfo) {
+		ci.PendingAmount = amt
+		ci.AddFundsMsg = &mcid
+	})
+
+	// Store a reference from the message CID to the channel, so that we can
+	// look up the channel from the message CID
+	err = ca.store.SaveNewMessage(channelInfo.ChannelID, mcid)
+	if err != nil {
+		log.Errorf("saving add funds message CID %s: %s", mcid, err)
+	}
+
+	go ca.waitForAddFundsMsg(from, to, mcid, cb)
+
+	return &mcid, nil
+}
+
+// waitForAddFundsMsg waits for mcid to appear on chain and returns error, if any
+func (ca *channelAccessor) waitForAddFundsMsg(from address.Address, to address.Address, mcid cid.Cid, cb onCompleteFn) {
+	err := ca.waitAddFundsMsg(from, to, mcid)
+	ca.msgWaitComplete(mcid, err, cb)
 }
 
-// WaitForAddFundsMsg waits for mcid to appear on chain and returns error, if any
-// TODO: wait outside the store lock!
-//  (tricky because we need to setup channel tracking before we know it's address)
-func (pm *Manager) waitForAddFundsMsg(ctx context.Context, mcid cid.Cid) {
-	defer pm.store.lk.Unlock()
-	mwait, err := pm.state.StateWaitMsg(ctx, mcid, build.MessageConfidence)
+func (ca *channelAccessor) waitAddFundsMsg(from address.Address, to address.Address, mcid cid.Cid) error {
+	mwait, err := ca.api.StateWaitMsg(ca.waitCtx, mcid, build.MessageConfidence)
 	if err != nil {
 		log.Error(err)
+		return err
 	}
 
 	if mwait.Receipt.ExitCode != 0 {
-		log.Errorf("voucher channel creation failed: adding funds (exit code %d)", mwait.Receipt.ExitCode)
+		err := xerrors.Errorf("voucher channel creation failed: adding funds (exit code %d)", mwait.Receipt.ExitCode)
+		log.Error(err)
+
+		ca.lk.Lock()
+		defer ca.lk.Unlock()
+
+		ca.mutateChannelInfo(from, to, func(channelInfo *ChannelInfo) {
+			channelInfo.PendingAmount = big.NewInt(0)
+			channelInfo.AddFundsMsg = nil
+		})
+
+		return err
 	}
-}
 
-func (pm *Manager) GetPaych(ctx context.Context, from, to address.Address, ensureFree types.BigInt) (address.Address, cid.Cid, error) {
-	pm.store.lk.Lock() // unlock only on err; wait funcs will defer unlock
-	var mcid cid.Cid
-	ch, err := pm.store.findChan(func(ci *ChannelInfo) bool {
-		if ci.Direction != DirOutbound {
-			return false
-		}
-		return ci.Control == from && ci.Target == to
+	ca.lk.Lock()
+	defer ca.lk.Unlock()
+
+	// Store updated amount
+	ca.mutateChannelInfo(from, to, func(channelInfo *ChannelInfo) {
+		channelInfo.Amount = types.BigAdd(channelInfo.Amount, channelInfo.PendingAmount)
+		channelInfo.PendingAmount = big.NewInt(0)
+		channelInfo.AddFundsMsg = nil
 	})
+
+	return nil
+}
+
+// Change the state of the channel in the store
+func (ca *channelAccessor) mutateChannelInfo(from address.Address, to address.Address, mutate func(*ChannelInfo)) {
```

Comment:
I think it may be worth looking at [go-statestore](github.com/filecoin-project/go-statestore) even if you're not using `go-statemachine` cause it provides some of this functionality in a simple form. (for the future)